### PR TITLE
Avoid unnecessary disk writes when saving user

### LINKED ISSUE #299: Reduce client save calls
It saves the user config whenever you connect to the network, change nick or join a channel. In these cases it should only call save if the thing it's trying to save isn't already there (e.g. join a channel, but its already in `join` in users config).

It currently uses a 1 second timer (a global time I might add) to throttle it.

### PR DESCRIPTION
**Pull request by xPaw**
Closes #299.

- Avoid unnecessary disk writes when saving user
- Makes writing user updates to disk asynchronous
- Replace ghetto timer with lodash's debounce

### CODE REVIEW

File: src/client.js [hunk 1]
```diff
@@ -292,19 +292,23 @@ Client.prototype.setPassword = function(hash, callback) {
...
-			client.config.password = hash;
-			callback(true);
-		} else {
-			callback(false);
-		}
+			// re-read the hash off disk to ensure we use whatever is saved. this will
+			// prevent situations where the password failed to save properly and so
+			// a restart of the server would forget the change and use the old
+			// password again.
+			var user = client.manager.readUserConfig(client.name);
```
Comment (xPaw) [inline src/client.js]: This is probably no longer necessary as the callback provides `err` from the write operation.

File: src/clientManager.js [hunk 2]
```diff
@@ -94,28 +94,34 @@ ClientManager.prototype.addUser = function(name, password) {
...
-		_.assign(user, opts);
-		fs.writeFileSync(
-			Helper.getUserConfigPath(name),
-			JSON.stringify(user, null, "\t")
-		);
-	} catch (e) {
-		log.error("Failed to update user", e);
-		return;
+	let user = this.readUserConfig(name);
+	const currentUser = JSON.stringify(user, null, "\t");
```
Comment (xPaw) [inline src/clientManager.js]: I timed this, and it's fast enough to do this (writes to disk are much slower, obviously). There's no really other easy way of comparing objects.

Comment (YaManicKill) [inline src/clientManager.js]: Yeah, this isn't ideal, and could cause some false positives (if the user's properties are now in a different order, but exactly the same) but I doubt that's going to cause any problems.

And it's definitely the fastest way of comparing them, it's just not the most accurate.

File: src/client.js (hunk 1 above)
Comment (YaManicKill) [inline src/client.js]: Yeah, this can be removed.

File: src/clientManager.js (hunk 2 above)
Comment (xPaw) [inline src/clientManager.js]: Actually order might cause issues due to sorting. Needs to be tested.

**Comment by xPaw** [comment]
Sounds like shouldn't be a problem.

File: src/client.js [hunk 3]
```diff
@@ -292,19 +292,23 @@ Client.prototype.setPassword = function(hash, callback) {
...
+			if (user.password === hash) {
+				client.config.password = hash;
+				callback(true);
+			} else {
+				callback(false);
```
Comment (astorije) [inline src/client.js]: That's quite the callback hell piling up here. Can we switch to using Promises for this? I believe that would make the code easier to follow :)

Comment (xPaw) [inline src/client.js]: A little out of scope of this PR.

**Comment by xPaw** [comment]
@YaManicKill, @astorije both comments addressed.

**Pull request merged**